File: backend/routes/order.go
package routes

import (
	"Project-M/config"
	"Project-M/internal/controller"

	"github.com/gin-gonic/gin"
)

func SetupOrderRoutes(v1 *gin.RouterGroup) {
	ctrl := controller.ProvideOrderController(config.DB())

	v1.POST("/orders", ctrl.CreateOrder)
	v1.GET("/orders", ctrl.ListOrders)
	v1.GET("/orders/:id", ctrl.GetOrder)
	v1.PATCH("/orders/:id", ctrl.UpdateOrder)
	v1.POST("/orders/:id/cancel", ctrl.CancelOrder)
	v1.POST("/orders/:id/close", ctrl.CloseOrder)

	v1.POST("/orders/:id/items", ctrl.AddItem)
	v1.PATCH("/orders/:id/items/:itemId", ctrl.UpdateItem)
	v1.DELETE("/orders/:id/items/:itemId", ctrl.DeleteItem)
	v1.PATCH("/orders/:id/items/:itemId/status", ctrl.UpdateItemStatus)
	v1.POST("/orders/:id/send-to-kitchen", ctrl.SendToKitchen)

	v1.GET("/kitchen/queue", ctrl.KitchenQueue)
}

File: frontend/src/lib/order.ts
import { apiClient } from "./apiClient";
import type { AddOrderItemInput, Bill, OpenOrderInput, Order, OrderItemStatus, OrderStatus } from "../types/order";

export type OrderListStatus = OrderStatus | "active" | "closed" | "";

export interface OrderListSummary {
  total: number;
  active: number;
  closed: number;
  statuses: Record<OrderStatus, number>;
}

export interface OrderListResponse {
  orders: Order[];
  pagination?: {
    page: number;
    limit: number;
    total: number;
    has_more: boolean;
  };
  summary?: OrderListSummary;
}

export const listOrders = (params?: {
  status?: OrderListStatus;
  payment_status?: "paid" | "unpaid" | "";
  search?: string;
  include_summary?: boolean;
  table_id?: number;
  date?: string;
  page?: number;
  limit?: number;
}) => apiClient.get<OrderListResponse>("/api/v1/orders", { params });

export const getOrder = (id: string | number) =>
  apiClient.get<Order>(`/api/v1/orders/${id}`);

export const createOrder = (data: OpenOrderInput) =>
  apiClient.post<Order>("/api/v1/orders", data);

export const updateOrder = (id: number, data: { customer_count: number; note?: string }) =>
  apiClient.patch<Order>(`/api/v1/orders/${id}`, data);

export const cancelOrder = (id: number, reason: string) =>
  apiClient.post<Order>(`/api/v1/orders/${id}/cancel`, { reason });

export const closeEmptyTableOrder = (id: number) =>
  apiClient.post<Order>(`/api/v1/orders/${id}/close-empty-table`);

export const closeOrder = (id: number) =>
  apiClient.post<Order>(`/api/v1/orders/${id}/close`);

export const getOrderBill = (id: number) =>
  apiClient.get<Bill>(`/api/v1/orders/${id}/bill`);

export const payOrder = (id: number, data: { method: "cash" | "promptpay_qr"; received_amount?: number; note?: string }) =>
  apiClient.post<Order>(`/api/v1/orders/${id}/pay`, data);

export const addOrderItem = (orderId: number, data: AddOrderItemInput) =>
  apiClient.post<Order>(`/api/v1/orders/${orderId}/items`, data);

export const updateOrderItem = (orderId: number, itemId: number, data: { quantity: number; note?: string }) =>
  apiClient.patch<Order>(`/api/v1/orders/${orderId}/items/${itemId}`, data);

export const deleteOrderItem = (orderId: number, itemId: number) =>
  apiClient.delete<Order>(`/api/v1/orders/${orderId}/items/${itemId}`);

export const updateOrderItemStatus = (orderId: number, itemId: number, status: OrderItemStatus, reason?: string) =>
  apiClient.patch<Order>(`/api/v1/orders/${orderId}/items/${itemId}/status`, { status, reason });

export const sendOrderToKitchen = (orderId: number) =>
  apiClient.post<Order>(`/api/v1/orders/${orderId}/send-to-kitchen`);

export const kitchenQueue = () =>
  apiClient.get<{ orders: Order[] }>("/api/v1/kitchen/queue");
